velox/expression/CastExpr.cpp

```diff
@@ -520,63 +520,80 @@ VectorPtr CastExpr::applyDecimal(
-/// @param context The context
-/// @param nullOnFailure Whether this is a cast or try_cast operation
-/// @param result The output vector
-template <bool castTo>
-void applyCustomTypeCast(
-    VectorPtr& input,
-    DecodedVector& inputDecoded,
-    const SelectivityVector& allRows,
-    const SelectivityVector& nonNullRows,
-    const CastOperatorPtr& castOperator,
-    const TypePtr& thisType,
-    const TypePtr& otherType,
+/// Apply casting from a non-custom type to a complex type.
+/// @param decoded The decoded input vector
+/// @param rows The selectivity vector of non-null rows in decoded.base().
+void CastExpr::applyComplexType(
+    const SelectivityVector& rows,
+    DecodedVector& decoded,
     exec::EvalCtx& context,
-    bool nullOnFailure,
+    const TypePtr& fromType,
+    const TypePtr& toType,
     VectorPtr& result) {
-  VELOX_CHECK_NE(
-      thisType,
-      otherType,
-      "Attempting to cast from {} to itself.",
-      thisType->toString());
-
-  exec::LocalSelectivityVector baseRows(
-      *context.execCtx(), inputDecoded.base()->size());
-  baseRows->clearAll();
-  context.applyToSelectedNoThrow(nonNullRows, [&](auto row) {
-    baseRows->setValid(inputDecoded.index(row), true);
-  });
-  baseRows->updateBounds();
+  switch (toType->kind()) {
+    // Handle complex type conversions
+    case TypeKind::MAP:
+      result = applyMap(
+          rows,
+          decoded.base()->asUnchecked<MapVector>(),
+          context,
+          fromType->asMap(),
+          toType->asMap());
+      break;
+    case TypeKind::ARRAY:
+      result = applyArray(
+          rows,
+          decoded.base()->asUnchecked<ArrayVector>(),
+          context,
+          fromType->asArray(),
+          toType->asArray());
+      break;
+    case TypeKind::ROW:
+      result = applyRow(
+          rows,
+          decoded.base()->asUnchecked<RowVector>(),
+          context,
+          fromType->asRow(),
+          toType->asRow());
+      break;
+    case TypeKind::SHORT_DECIMAL:
+    case TypeKind::LONG_DECIMAL:
+      result = applyDecimal(rows, decoded, context, fromType, toType);
+      break;
+    default: {
+      VELOX_UNREACHABLE();
+    }
+  }
+}
 
-  VectorPtr localResult;
-  if constexpr (castTo) {
-    context.ensureWritable(*baseRows, thisType, localResult);
+/// Apply casting between a custom type and another type, or to a complex type.
+/// @param decoded The decoded input vector
+/// @param rows The selectivity vector of non-null rows in decoded.base().
+void CastExpr::applyComplexAndCustomType(
+    const SelectivityVector& rows,
+    DecodedVector& decoded,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType,
+    VectorPtr& result) {
+  if (castFromOperator_ || castToOperator_) {
+    VELOX_CHECK_NE(
+        fromType,
+        toType,
+        "Attempting to cast from {} to itself.",
+        fromType->toString());
+    context.ensureWritable(rows, toType, result);
```

Comment:
It would be nice to let the castToOperator decide whether they want to make a new 'result' vector or reuse the one provided. In other words, I think we should remove this call here and let castTo handle the 'result' similar to how VectorFunction::apply does.